### PR TITLE
Integration Tests using Maven-invoker-plugin

### LINKED ISSUE #1345: Enforcer Rule Integration Test?
How can we ensure the enforcer rule works fine as part of precommit checks?

# Consideration

The check needs to pick up right version of linkage-checker-enforcer-rule.

The check needs to have the enforcer rule plugin to read the XML configuration to detect errors to detect errors like https://github.com/GoogleCloudPlatform/cloud-opensource-java/issues/1343.

# Options

## Directories for Integration Tests

- `enforcer-rules/integration-test/project-with-exclusion-file/pom.xml`
- `enforcer-rules/integration-test/project-with-report-only-reachable/pom.xml`
...

and a shell script to `cd` these directories and run `mvn verify`?

## Maven's way to run integration tests

https://cwiki.apache.org/confluence/display/MAVEN/Creating+a+Maven+Integration+Test. This requires `AbstractMavenIntegrationTestCase` in Maven's [core-integrating-testing module](https://github.com/apache/maven-integration-testing/blob/master/core-it-support/maven-it-helper/src/main/java/org/apache/maven/it/AbstractMavenIntegrationTestCase.java). As we're not testing Maven itself, the class might not be suitable for our use case.

## Maven checkstyle-plugin's tests

Checkstyle plugin is good candidate to follow the testing setup.

https://github.com/apache/maven-checkstyle-plugin/tree/master/src/it/MCHECKSTYLE-338

```
          <artifactId>maven-checkstyle-plugin</artifactId>
          <version>@pom.version@</version>
```

Checkstyle test has verify.groovy

It has "run-its" profile to execute [maven-invoker-plugin](http://maven.apache.org/plugins/maven-invoker-plugin/):

```
  <profiles>
    <profile>
      <id>run-its</id>
      <build>
        <plugins>
          <plugin>
            <artifactId>maven-invoker-plugin</artifactId>
            <configuration>
              <goals>
                <goal>clean</goal>
                <goal>site</goal>
...
```

> The Invoker Plugin is used to run a set of Maven projects. The plugin can determine whether each project execution is successful, and optionally can verify the output generated from a given project execution.

This plugin is in particular handy to perform integration tests for other Maven plugins. The Invoker Plugin can be employed to run a set of test projects that have been designed to assert certain features of the plugin under test.

### PR DESCRIPTION
**Pull request by suztomo**
Fixes #1345 

[maven-invoker-plugin](http://maven.apache.org/plugins/maven-invoker-plugin/index.html):

> This plugin is in particular handy to perform integration tests for other Maven plugins. 

CircleCI runs the integration tests successfully ([link](https://circleci.com/gh/GoogleCloudPlatform/cloud-opensource-java/3816?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)):

```
[INFO] --- maven-invoker-plugin:3.2.1:run (integration-test) @ linkage-checker-enforcer-rules ---
[INFO] Building: fail-build-for-linkage-errors/pom.xml
[INFO] run post-build script verify.groovy
[INFO]           fail-build-for-linkage-errors/pom.xml ............ SUCCESS (47.1 s)
[INFO] Building: report-only-reachable/pom.xml
[INFO] run post-build script verify.groovy
[INFO]           report-only-reachable/pom.xml .................... SUCCESS (4.5 s)
[INFO] -------------------------------------------------
[INFO] Build Summary:
[INFO]   Passed: 2, Failed: 0, Errors: 0, Skipped: 0
[INFO] -------------------------------------------------
```

# verify.groovy

Maven-invoker-plugin has two ways to verify the build result: BeanShell and Groovy. As we'll be working with Gradle plugins soon, I choose Groovy that powers build.gradle files.

# mvnw

Kokoro build has outdated Maven version (3.3.9), which prevents the maven-invoker-plugin integration tests, causing diamond dependency conflicts with aether classes  (#1361 ). To solve this, this PR uses https://github.com/takari/maven-wrapper to have consistent Maven version across CI servers. I referenced [spring-cloud-gcp](https://github.com/spring-cloud/spring-cloud-gcp) for mvnw usage.

### CODE REVIEW

File: enforcer-rules/pom.xml [hunk 1]
```diff
@@ -128,4 +128,27 @@
...
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-invoker-plugin</artifactId>
+        <version>3.2.1</version>
+        <configuration>
+          <cloneProjectsTo>${project.build.directory}/it</cloneProjectsTo>
+          <localRepositoryPath>${project.build.directory}/local-repo</localRepositoryPath>
+          <postBuildHookScript>verify</postBuildHookScript>
```
Comment (suztomo) [inline enforcer-rules/pom.xml:140]: Document for the parameters: http://maven.apache.org/plugins/maven-invoker-plugin/run-mojo.html#cloneProjectsTo

File: enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy [hunk 2]
```diff
@@ -0,0 +1,8 @@
+def buildLog = new File(basedir, "build.log")
```
Comment (suztomo) [inline enforcer-rules/src/it/fail-build-for-linkage-errors/verify.groovy:1]: This is groovy script specified by ["postBuildHookScript"](http://maven.apache.org/plugins/maven-invoker-plugin/run-mojo.html#postBuildHookScript).

Maven-checkstyle-plugin also uses groovy script to verify the build https://github.com/apache/maven-checkstyle-plugin/tree/master/src/it/MCHECKSTYLE-338

**Comment by suztomo** [comment]
The integration tests failed unexpectedly in Windows ([log](https://fusion.corp.google.com/runanalysis/buildlogs/prod%3Acloud-opensource-java%2Fgcp_windows%2Fpresubmit/prod%3Acloud-opensource-java%2Fgcp_windows%2Fpresubmit/KOKORO/b2ec87ca-a246-4e68-b17e-2409870e2037/0/prod%3Acloud-opensource-java%2Fgcp_windows%2Fpresubmit/Targets)):

```
       |        [INFO] ------------------------------------------------------------------------
       |        [INFO] BUILD FAILURE
       |        [INFO] ------------------------------------------------------------------------
       |        [INFO] Total time: 14.093 s
       |        [INFO] Finished at: 2020-04-17T12:14:40-07:00
       |        [INFO] Final Memory: 19M/285M
       |        [INFO] ------------------------------------------------------------------------
       |        [ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce (enforce-linkage-checker) on project no-such-method-error-example: Execution enforce-linkage-checker of goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce failed: An API incompatibility was encountered while executing org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3:enforce: java.lang.NoSuchMethodError: org.eclipse.aether.transfer.TransferResource.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/io/File;Lorg/eclipse/aether/RequestTrace;)V
       |        [ERROR] -----------------------------------------------------
       |        [ERROR] realm =    plugin>org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M3
       |        [ERROR] strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
       |        [ERROR] urls[0] = file:/T:/src/github/cloud-opensource-java/enforcer-rules/target/local-repo/org/apache/maven/plugins/maven-enforcer-plugin/3.0.0-M3/maven-enforcer-plugin-3.0.0-M3.jar
       |        [ERROR] urls[1] = file:/T:/src/github/cloud-opensource-java/enforcer-rules/target/local-repo/com/google/cloud/tools/linkage-checker-enforcer-rules/1.2.2-SNAPSHOT/linkage-checker-enforcer-rules-1.2.2-SNAPSHOT.jar
```

File: enforcer-rules/pom.xml [hunk 3]
```diff
@@ -105,7 +105,6 @@
     <dependency>
       <groupId>org.apache.maven.resolver</groupId>
       <artifactId>maven-resolver-api</artifactId>
-      <scope>provided</scope>
```
Comment (suztomo) [inline enforcer-rules/pom.xml:108]: Given maven-resolver-api was at the end of the class path, I thought controlling the order of the dependency would solve the issue.

It did not help. https://gist.github.com/suztomo/d35ffca4e31387591a3f7b18b7366a2b

File: .mvn/wrapper/MavenWrapperDownloader.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (suztomo) [inline .mvn/wrapper/MavenWrapperDownloader.java:1]: This file is from `mvn -N io.takari:maven:wrapper` https://github.com/takari/maven-wrapper.

File: .mvn/wrapper/maven-wrapper.properties [hunk 5]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.3/apache-maven-3.6.3-bin.zip
```
Comment (suztomo) [inline .mvn/wrapper/maven-wrapper.properties:1]: This file is from `mvn -N io.takari:maven:wrapper` https://github.com/takari/maven-wrapper.

File: README.md [hunk 6]
```diff
@@ -35,7 +35,7 @@ This project is built using _Maven_.
 
 ## Requirements
 
-1. Maven 3.5.0 or later.
+1. Maven 3.6.0 or later.
```
Comment (suztomo) [inline README.md:38]: I found Maven 3.5.0 does not work for the enforcer rule because of the dependency conflict #1361

File: mvnw [hunk 7]
```diff
@@ -0,0 +1,310 @@
+#!/bin/sh
```
Comment (suztomo) [inline mvnw:1]: This file is from `mvn -N io.takari:maven:wrapper` https://github.com/takari/maven-wrapper.

File: mvnw.cmd [hunk 8]
```diff
@@ -0,0 +1,182 @@
+@REM ----------------------------------------------------------------------------
```
Comment (suztomo) [inline mvnw.cmd:1]: This file is from `mvn -N io.takari:maven:wrapper` https://github.com/takari/maven-wrapper.

File: .mvn/wrapper/MavenWrapperDownloader.java (hunk 4 above)
Comment (elharo) [reply .mvn/wrapper/MavenWrapperDownloader.java:1]: I'm surprised this isn't available as a jar on maven central.

**Pull request merged**